### PR TITLE
Add `--python-platform` to `sync` and `install` commands

### LINKED ISSUE #2079: Cross-platform support
My goal is to support building OCI (docker) images for a platform other than the host platform.

For reasons related to our tooling (we use Bazel and want hermetic builds), we do not want to do this via running `pip install` commands on the image itself (as would be normal in a Dockerfile).

The solution we have now is to download wheels for different platforms using `pip download --platform`. We can then pass the appropriate target depending on which platform we are building for. This method has the downside that we need appropriate wheels to exist for the packages we want to use (this sometimes involves creating our own).

Is there any way to achieve my goal with uv? Support for `uv pip download --platform` would be sufficient, but it would be ideal to be able to not need the wheels. `uv pip install --platform` or similar would be even better (no idea if this is at all feasible).

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

pip supports providing a `--platform` to `pip install`, which can be used to seed an environment (e.g., for use in a container or otherwise). This PR adds `--python-platform` to our commands to support a similar workflow. It has some caveats, which are documented on the CLI.

Closes #2079.

### CODE REVIEW

File: crates/uv/src/cli.rs [hunk 1]
```diff
@@ -834,6 +834,27 @@ pub(crate) struct PipSyncArgs {
...
+    /// The platform for which requirements should be installed.
+    ///
+    /// Represented as a "target triple", a string that describes the target platform in terms of
+    /// its CPU, vendor, and operating system name, like `x86_64-unknown-linux-gnu` or
+    /// `aaarch64-apple-darwin`.
+    ///
+    /// WARNING: When specified, uv will select wheels that are compatible with the target platform.
+    /// The resulting environment may not be fully compatible with the current platform. Further,
+    /// distributions that are built from source may ultimately be incompatible with the target
+    /// platform. This option is intended for cross-compilation and other advanced use cases.
```
Comment (BurntSushi) [inline crates/uv/src/cli.rs:854]: I think this was mentioned in #3111, but how can we provide end users with a list of valid triples they can use? I don't think we should just link to the Rust docs unless we specifically want to guarantee a correspondence there. Basically, my concern here is that users won't know what they're allowed to put here.

Honestly, I have a similar concern with pip's `--platform` flag. Its docs are:

```
  --platform <platform>       Only use wheels compatible with <platform>.
                              Defaults to the platform of the running system.
                              Use this option multiple times to specify
                              multiple platforms supported by the target
                              interpreter.
```

After reading that, I have no clue whatsoever what I'm allowed or even supposed to use as a value for this flag.

(And now I'm wondering what the difference is between pip's `--platform` flag and our `--python-platform` flag...)

**Review by BurntSushi — APPROVED** [review]
I was having deja vu here, because I thought I had reviewed a change like this already. I did some digging and:

* #3111 added initial support. It added it as `--platform` to _only_ `uv pip compile`.
* #3146 renamed the flag from `--platform` to `--python-platform`, since `pip` also has a `--platform` flag and it doesn't match how ours behaves.
* #3147 adds `--python-platform` to the configuration.
* _This_ PR adds `--python-platform` to the `uv pip sync` and `uv pip install` commands.

So with all that context, I think this change makes sense, but:

* Can we add a test covering this flag for the `uv pip sync` and `uv pip install` commands? I think we have a test for `uv pip compile`, but it would be good to cover the other two as well.
* I mentioned this in a review comment, but what is the difference between pip's `--platform` and this `--python-platform` flag? And does there need to be a difference? Can we match what `pip` does instead?

**Review by konstin — APPROVED** [review]

File: crates/uv/src/cli.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv/src/cli.rs:854]: We already do! Clap shows all the variants automatically along with their Rustdoc.

**Comment by zanieb** [comment]
> Can we add a test covering this flag for the uv pip sync and uv pip install commands? I think we have a test for uv pip compile, but it would be good to cover the other two as well.

Fwiw we've been intentionally _not_ covering flags across every command. No comment on whether or not that's best.

File: crates/uv/src/cli.rs (hunk 1 above)
Comment (BurntSushi) [reply crates/uv/src/cli.rs:854]: Oh sweet. Perfect.

**Comment by charliermarsh** [comment]
(I'll answer your other question after goal setting, it's a good one to record in GitHub.)

**Comment by charliermarsh** [comment]
> ...but what is the difference between pip's --platform and this --python-platform flag? And does there need to be a difference? Can we match what pip does instead?

Okay, so pip supports `--platform`, `--abi`, and `--python-version`. When provided, it basically takes those, and generates all the compatible target tags by generating combinations of the tree. If you provide `--platform`, you _must_ either run with `--no-deps` or `--only-binary :all:`. And as far as I can tell, pip's `--platform` does not affect _resolution_ (i.e., it does not change the markers; only the wheel tags).

So our `--python-platform` argument is slightly different. It's "higher-level", in that you describe a platform and then we generate the relevant wheel tags and markers. And we ship with fewer limitations, but with the risk that users may misuse the flag and create broken environments.

File: crates/uv/src/cli.rs [hunk 2]
```diff
@@ -846,6 +846,27 @@ pub(crate) struct PipSyncArgs {
...
```
Comment (danielhollas) [inline crates/uv/src/cli.rs:866]: This last sentence is confusing given the preceding paragraph, perhaps you meant

```suggestion
    /// platform. This option is not intended for cross-compilation and other advanced use cases.
```
?

Comment (charliermarsh) [reply crates/uv/src/cli.rs:866]: I think "cross-compilation" is the wrong term. It _is_ intended for advanced use-cases, but "cross-compilation" emphasizes building from source in the wrong way.

Comment (danielhollas) [reply crates/uv/src/cli.rs:866]: Ah, okay. Yeah, I was definitely also mislead by the preceding sentence that specifically talks about building from source.

**Pull request merged**